### DISCUSSION TITLE
PdfReader - Extract images from specific pages

### DISCUSSION THREAD

**Discussion by FrsECM** [py-pdf/pypdf#2536] [Q&A] [answered]
Replace this: What happened? What were you trying to achieve?

## Environment
Python 3.8
WSL Ubuntu 22.04
Windows11
pypdf  4.1.0

## Issue
I generated a very simple pdf with libreoffice-writer :
[test_image.pdf](https://github.com/py-pdf/pypdf/files/14733274/test_image.pdf)

In this pdf, there is two pages, one containing a small text, another containing an image.
<img src='https://github.com/py-pdf/pypdf/assets/26071804/67f68340-a8a0-4b7d-aca7-17f243f7e4fe' height='200' />
<img src='https://github.com/py-pdf/pypdf/assets/26071804/0b80958a-7d76-40c3-a3a1-4795983b3039' height='200' />

I want to extract pdf pages and get the image only in the second page.

The code to reproduce the issue is here : 
```python
pdf = PdfReader(pdf_file)
for i,page in enumerate(pdf.pages):
        print(f'--- Extracting page {i}')
        print(page.extract_text())
        print(len(page.images))
```
The result is bellow : 
```
--- Extracting page 0
Test page 1
1
--- Extracting page 1
1
```
I expect that on page 0 there is 0 image in order to extract the image only from the second page.
I don't know if it is a normal behaviour.

**How to do what i would like to obtain ?**
Thanks,
Regards

**Comment by pubpub-zz** [comment, marked as answer]
Your pdf has the image attached to both pages:
![image](https://github.com/py-pdf/pypdf/assets/4083478/3de4f489-6a72-4f73-bc60-22c58b47b344)

pypdf do not check if the images are "called" in the image content.

**Comment by FrsECM** [comment]
I don’t know well how works pdf under the hood.
**What is the name of the GUI tool you are using ?**

**Can we get this ownership information from somewhere when we create the « PageObject » ?**

**Reply by stefan6419846** [reply]
The shown tool is Apache PDFBox Debugger standalone: https://pdfbox.apache.org/download.html

> Can we get this ownership information from somewhere when we create the « PageObject » ?

What do you mean by this? You already have the ownership information if you use `page.images`.

**Reply by pubpub-zz** [reply]
the tool I've used is PdfBox in debug mode: https://pdfbox.apache.org/download.cgi
if you mean to know if the Page content the image. You have to look in the contents for an operation Do where the parameter is the key of the resource (here /Im7)

**Comment by FrsECM** [comment]
I mean that the ressources seems to be shared between all Pages like if it was a reference. Ownership may not be the right term.

I saw with a GUI tool that in the first page, there is nothing in the stream concerning the image, there is only BT and ET, corresponding to the text.
![image](https://github.com/py-pdf/pypdf/assets/26071804/4146a70d-64c6-44b7-b1c4-621e25766451)

But on the second one, we can see in the stream that there is the image :
![image](https://github.com/py-pdf/pypdf/assets/26071804/a94c74d1-9241-42c4-91ff-7acdbfe7b181)

**Maybe we can use the content to confirm if the image is owned by the PageObject in that case ?**
![image](https://github.com/py-pdf/pypdf/assets/26071804/c941452e-c1cc-4f84-b57f-25ef3dc67a0e)

It would allow to filter the **images** property before returning it.

**Reply by FrsECM** [reply]
> the tool I've used is PdfBox in debug mode: https://pdfbox.apache.org/download.cgi if you mean to know if the Page content the image. You have to look in the contents for an operation Do where the parameter is the key of the resource (here /Im7)

I saw your answer just after my publication.

**Do you think it is releavant to modify the default behaviour of the library regarding that point ?**

I can do a proposal regarding that point in a PR.

**Reply by pubpub-zz** [reply]
> I saw your answer just after my publication.
> 
> **Do you think it is releavant to modify the default behaviour of the library regarding that point ?**

I would not modify the current extraction as there could be other reasons for an image to not be displayed(already met): If an image is out of the page it will be present in the `get_contents()`

You can propose a function in pages `is_image_displayed(image_id:str)`. at least consider:
* images called in contents or in XObjects/Forms within the page
* image intersect with the cropped area

**Reply by andrea-santoro1-a2a** [reply]
hi, has this been solved somehow?

**Reply by stefan6419846** [reply]
No, there has been no related upstream change as far as I am aware. If you want to work on this, consider opening an issue outlining your proposed solution before starting with a PR.

**Reply by andreasntr** [reply]
@stefan6419846 as a temporary workaround, i've implemented something similar for pymupdf today (because it was easier): https://github.com/pymupdf/PyMuPDF/pull/4961 . Basically I'm keeping only images whose actual references (xrefs) are in a given page.

I'm having issues understanding how to extract xrefs in pypdf to implement something similar, but I'm willing to. If you can guide me I can try posting a PR

**Reply by stefan6419846** [reply]
> Basically I'm keeping only images whose actual references (xrefs) are in a given page. [...] If you can guide me I can try posting a PR

As per my previous comment, please outline your proposed solution in an issue first to properly discuss it with keeping backwards-compatibility in mind.

> I'm having issues understanding how to extract xrefs in pypdf to implement something similar, but I'm willing to.

Please be more specific in the corresponding issue where you see problems at the moment. The `ImageFile` class knows its corresponding objects, arbitrary objects can be retrieved from the associated `PdfDocCommon` instance.
